Frame-based profiling and external mode

ShadauxCat opened the pull request:
Added support for frame-based profiling mode as well as support for running perf_timer as an external viewer application for metrics gathered in other programs written in other languages.

ShadauxCat committed:
Added support for frame-based profiling mode as well as support for running perf_timer as an external viewer application for metrics gathered in other programs written in other languages.

ShadauxCat committed:
reverting a change that was made in error

zoejbare commented on README.md:
```diff
@@ -8,21 +30,166 @@ To enable:
+
+If you're profiling a large number of frames and having difficulty clicking on the frame you want, you can also click and drag within the header to zoom in on a segment of the output so that you can click on the one you want more easily.
+
+####!! A word of warning !!
```
Need a space between the `####` and the `!!` for the formatting to work.

zoejbare commented on README.md:
```diff
@@ -8,21 +30,166 @@ To enable:
+
+The binary format is similar to the JSON format, but, perhaps expectedly, is more strict. It's provided to allow for native applications to generate the output more efficiently.
+
+**Note that the binary format does not include ANY alignment padding to ensure it will work across all languages and architectures without making any assumptions about the alignment of data. Structs used to generate the metrics file should thus be packed. It is, however, assumed the output will be generated on the same system as the metrics file, so native endianness is used and it is not necessa ...
```
While I think it's fine to have it assume only native endianness for now, I think it'll be important eventually to have some detection and handling for both big and little endian data.  For instance, it's very possible for someone to generate a metrics file on a big endian system, then generate the report for it on a little endian system.

ShadauxCat replied:
But... why would they? It's easy on the python side to force endianness, but it adds a lot of implementation difficulty on the side of the application generating the metrics.

ShadauxCat replied:
well, set it to little endian anyway. Won't make a difference on intel systems so I guess it's only overhead if people are on ARM.

zoejbare replied:
A user might be generating metrics on a console platform that happens to be big endian.

zoejbare replied:
Forcing the format to little endian is a fine way of handling it.

zoejbare commented on README.md:
```diff
@@ -8,21 +30,166 @@ To enable:
+
+| Field | Type | Size | Values |
+|-------|------|------|--------|
+| Magic | unsigned int32 |4 bytes | MUST contain the value 0xFA57 |
```
Is it intentional that this field is a 32-bit uint, but the required value is a 16-bit uint?  If so, that's fine, I just got confused when I looked at it because it means half the magic number would always be zero, which is valid.

ShadauxCat replied:
Yeah, I'm just not fond of using 16-bit values for magic. 32-bit magic seems at least reasonably standardized.

ShadauxCat committed:
Fixed a readme problem and forced endianness to little

zoejbare commented:
Are 1000 sample files in the repo really necessary?  We could probably have gotten the point across with 10, and the built in test can still generate 1000 on its own.

ShadauxCat commented:
> Are 1000 sample files in the repo really necessary? We could probably have gotten the point across with 10, and the built in test can still generate 1000 on its own.

There are 10 sample files as well. I wanted to provide something users could quickly demo the application with that would show more than 10 samples worth of functionality.

zoejbare commented:
Then, how about 100?  That's a reasonable sample size and an order of magnitude fewer files.

zoejbare reviewed (approved):
Overall, it looks fine to me.

zoejbare commented on perf_timer.py:
```diff
@@ -476,11 +488,11 @@ def __exit__(self, excType, excVal, excTb):
 
 	@staticmethod
-	def PrintPerfReport(reportMode, output=None):
+	def PrintPerfReport(reportMode, output=None, name=None):
```
Just noticed this function's docstring has not been updated with the new parameter.

zoejbare replied:
Also, in the doc for the `output` parameter, the word "tree" is missing a double quote.  And I'm not sure if the `output` doc is still correct with the addition of `name`.

zoejbare commented:
Upon inspecting it, I think the `samples` directory is misleading.  I would expect it to have samples showing how to use the perf timer, but it's really just sample output.  Instead, we have no sample usage at all.

zoejbare commented on perf_timer.py:
```diff
@@ -492,6 +504,170 @@ def PrintPerfReport(reportMode, output=None):
+				break
+
+		longest = 0
+		if len(elementsByFrame) != 1 and reportMode == ReportMode.HTML:
```
We should really have a check before this `if` statement to check for `elementsByFrame == 0`.  If you attempt to output a report with no perf data, it won't break, but the html output is somewhat invalid.

ShadauxCat committed:
- Added support for a "note" type of event
- Changed block names in binary format from null-terminated to having their length encoded with them (it's only one byte more and much faster to read)
- Added a reference C++ header implementation

ShadauxCat committed:
Fixed an issue with threaded output, addressed review feedback.

ShadauxCat committed:
Addressed more review feedback.

ShadauxCat commented:
all review feedback has been addressed

zoejbare commented on lib_reference/c++/PerfTimer.hpp:
```diff
@@ -0,0 +1,340 @@
+				perror("Could not open perf_timer output file for writing.");
+				return;
+			}
+			int magic = 0xFA57;
```
Should this not be something like `int32_t` to guarantee size?

zoejbare commented on perf_timer.py:
```diff
@@ -476,22 +503,204 @@ def __exit__(self, excType, excVal, excTb):
+					output("Frame #{}".format(key))
+					output("==============================")
+				elif __name__ == "__main__":
+					sys.stdout.write("\rGenerating individual frame output for frame {}...".format(key))
```
You have `\r` starting off this string.  Did you mean `\n`?

ShadauxCat replied:
I did not. `\r` is carriage return - return to beginning of line. `\n` moves the carat to the next line. This is overwriting the existing line so that only the number at the end changes and it doesn't spam the screen.

ShadauxCat committed:
Switching some values to explicit sizes.
